### ISSUE TITLE
[Tracking issue] Rearrange location of some of the existing lints

### ISSUE THREAD

**Issue by magurotuna** [rust-lang/rust-clippy#6680]
Currently, the lint rules grouped as "methods" reside in `clippy_lints/src/methods` as there are a number of rules belonging to the "methods" group.
However, although the other groups such as "loops", "types", "transmute" and "misc" also have many relavent rules, they reside not in `clippy_lints/src/<group_name>` but in `clippy_lints/src`.
It would be great to have these groups located under `<group_name>` directory just like "methods". So I open this issue to do this refactoring and to keep track of it.

- [ ] loops
- [ ] types
- [ ] transmute
- [ ] misc

How to rearrange is like the following, proposed by @flip1995 on [zulip](https://rust-lang.zulipchat.com/#narrow/stream/257328-clippy/topic/lint.20file.20location.20clippy.236643/near/225318775)

> I would keep the declare_clippy_lint! in the mod.rs Best case would be that mod.rs only contains the lint definitions and the LintPass impl. Utility functions of the module should probably go in methods/utils.rs and then reexported for the module in mod.rs.

CC @nahuakang @camsteffen

**Comment by nahuakang** [comment]
@magurotuna @camsteffen Shall we kick this off sometime this month? I could focus on `loops` since I just came from it (and there are only 18 lints). But I think for `methods` we'd need to split carefully since it has 50+ lints :thinking:

**Comment by flip1995** [comment]
Also feel free to split this not only in multiple commits, but also in multiple PRs.

**Comment by camsteffen** [comment]
@nahuakang Sure! Frankly I don't want to promise anything from me at this point. For organization, I would recommend we comment here on what group you're currently working on. But you don't have to do an entire group.

**Comment by phansch** [comment]
> clippy_lints/src/<group_name>/<lint_name>.rs

If `group_name` refers to lint groups like `restriction` or `pedantic`, it's worth noting that these change quite often compared to lint names. So, if someone changes the lint group of a lint they will have to move it to the correct directory as well, adding some extra work. Not sure if that's fine?

If it just refers to unrelated group names, ignore what I said =)

**Comment by nahuakang** [comment]
@phansch The `group_name` would be `methods` or `loops` so that new contributors have an easier time finding lint's or understanding the structure of the repo 😁 You are right that it shouldn't be the lint group categories.

**Comment by magurotuna** [comment]
I'll work on "transmute" group :)

**Comment by Y-Nak** [comment]
I'll work on `types`!

**Comment by flip1995** [comment]
@Y-Nak @nahuakang @magurotuna Do you want to cross-review your PRs to make sure that you all used the same "style" for the restructuring, so the Clippy code is consistent? That would be a big help for me reviewing those PRs because I don't have to compare them to eachother.

**Comment by nahuakang** [comment]
Can do. Mine is ready for review. @Y-Nak @magurotuna is either of yours ready for a first comparison? :)

**Comment by flip1995** [comment]
After a quick scan of the PRs it seems that all are ready for review. 👍

**Comment by magurotuna** [comment]
Yes, mine is ready too. I'll have a look at the PRs of @nahuakang and @Y-Nak when I have time :)

**Comment by Y-Nak** [comment]
@flip1995 Sure, I'll review them.
@nahuakang @magurotuna mine is also ready, please see #6724 before reviewing my PR.

**Comment by Y-Nak** [comment]
`function.rs` also can be refactored, I think.

**Comment by magurotuna** [comment]
@Y-Nak Good catch! I updated the PR description.

**Comment by TaKO8Ki** [comment]
Can I work on `methods `?

**Comment by magurotuna** [comment]
@TaKO8Ki
I think you may go ahead! Updated PR description.

**Comment by camsteffen** [comment]
Some good teamwork happening in here!

**Comment by TaKO8Ki** [comment]
@magurotuna 
I decided to split into a new PR because https://github.com/rust-lang/rust-clippy/pull/6826 became too large. So, could you add a link to my tracing issue https://github.com/rust-lang/rust-clippy/issues/6886 to this issue?

**Comment by magurotuna** [comment]
@TaKO8Ki Sure thing! Done it.

**Comment by Y-Nak** [comment]
I'm working on `function`.

Btw, it would be better to discuss how to refactor `misc`. All lints in `misc` seem not relevant to each other at all, so I think it makes sense to declare lint passes for each lint.

**Comment by magurotuna** [comment]
> Btw, it would be better to discuss how to refactor misc. All lints in misc seem not relevant to each other at all, so I think it makes sense to declare lint passes for each lint.

Agreed. I don't see any specific reason why the lints that are currently categorized as "misc" are there. "misc" is so ambiguous that it makes people think a new lint that seems trivial _to them_ should go into "misc". What kind of lints seem trivial to people differs from person to person IMO. Thus it makes more sense to me too, to separate the lints in "misc" and declare individual lint passes for them.

**Comment by flip1995** [comment]
I think for `misc` some of the lints there would also fit in already existing modules. `misc` and `misc_early` are both artifacts from the early days of Clippy. I agree that it might be good to split up and get rid of those modules completely.

**Comment by magurotuna** [comment]
Cool! I updated the description of this issue, adding a `misc_early` checkbox and how to refactor `misc` and `misc_early`.
I would like to work on `misc` :)

**Comment by Y-Nak** [comment]
@magurotuna Could you update the TODO list? Refactoring `functions` has been done in https://github.com/rust-lang/rust-clippy/pull/6990.

**Comment by magurotuna** [comment]
@Y-Nak Great job! Updated.

**Comment by Y-Nak** [comment]
Oh, we missed `matches.rs`.
@magurotuna could you add it to the TODO list, please?

**Comment by magurotuna** [comment]
@Y-Nak Sure, added!

----
Anyway, recently I've been mostly working on another project so I don't have enough time to refactor `misc` despite of my previous post. I'll withdraw from `misc` work for the moment. If anyone is interested in it, please feel free to take it over :D

**Comment by TaKO8Ki** [comment]
Can I also work on `misc_early` group?

**Comment by flip1995** [comment]
Sure! If you have questions on where lints should go, I think the best place to discuss this is on Zulip.

**Comment by ada4a** [comment]
I think `matches` can be crossed out as well, thanks to https://github.com/rust-lang/rust-clippy/pull/8400
